### PR TITLE
KAZOO-5878: Add parameters stability level checks for crossbar request objects

### PR DESCRIPTION
**Pull request by harenson**

### CODE REVIEW

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 1]
```diff
@@ -115,3 +116,38 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+
+maybe_check_params_stability_level(Value, Schema, State) ->
+    case kz_json:get_value(<<"properties">>, Schema) of
+        'undefined' -> State; %% Skipping schema with out `properties' property
+        Properties -> check_params_stability_level(Properties, Value, State)
+    end.
+
+check_params_stability_level(Properties, Value, State) ->
+    {ok, CBConfig} = kz_datamgr:open_cache_doc(<<"system_config">>, <<"crossbar">>),
+    SystemSL = kz_json:get_binary_value([<<"default">>, <<"stability_level">>], CBConfig),
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: To get a value from system_config you have to use `kapps_config` module.

So instead of this line and above line use this:
```
kapps_config:get_ne_binary_value(<<"crossbar">>, <<"stability_level">>, <<"stable">>)
```
It gets `stability_level` key from crossbar category (a document in system_config db), if it's not found it set/return the default value `stable`.

After this change make sure to compile and run `make apis` to update crossbar system_config schema.

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 2]
```diff
@@ -115,3 +116,38 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+    SystemSLInt = stability_level_to_int(SystemSL),
+    Fun = fun(Key, _Val, Acc) ->
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: move this to it's own module function.

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 3]
```diff
@@ -115,3 +116,38 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+              SearchKey = [Key, <<"stability_level">>],
+              PropSL = kz_json:get_value(SearchKey, Properties, <<"stable">>),
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: make inline

```
PropSL = stability_level_to_int(kz_json:get_value([Key, <<"stability_level">>], Properties, <<"stable">>)),
```

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 4]
```diff
@@ -115,3 +116,38 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+              case  stability_level_to_int(PropSL) < SystemSLInt of
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]:
```
case  PropSL =:= SystemSLInt
    orelse  PropSL < SystemSLInt
of

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 5]
```diff
@@ -115,3 +116,43 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
             jesse_error:handle_data_invalid('external_error', ErrorMsg, State)
     end.
+
+maybe_check_params_stability_level(Value, Schema, State) ->
+    case {kz_json:get_value(<<"properties">>, Schema),
+          kz_json:get_value(<<"id">>, Schema),
+          kz_json:is_json_object(Value)}of
+        {'undefined', _, _} ->
+            State; %% Skipping schema with out `properties' property
+        {Properties, Id, true} when Id /= <<"system_config.crossbar">> ->
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: Why?

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 6]
```diff
@@ -115,3 +116,43 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+            check_params_stability_level(Properties, Value, State);
+        _ ->
+            State
+    end.
+
+check_params_stability_level(Properties, Value, State) ->
+    SystemSL = kapps_config:get_binary(<<"crossbar">>, <<"stability_level">>, <<"stable">>),
+    SystemSLInt = stability_level_to_int(SystemSL),
+    Fun = fun(Key, _Val, Acc) ->
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: add a new function:

```erlang
-spec check_param_stability(kz_json:key(), kz_json:json_term(), kz_json:keys()) -> kz_json:keys().
check_param_stability(Key, _Val, Acc) ->
    SearchKey = [Key, <<"stability_level">>],
    PropSL = kz_json:get_value(SearchKey, Properties, <<"stable">>),

    case stability_level_to_int(PropSL) < SystemSLInt of
        true -> [Key | Acc];
        false -> Acc
    end.
```

and call it `kz_json:foldl(fun check_param_stability/3, [], Value)`.

File: core/kazoo_schemas/src/kz_json_schema.erl [hunk 7]
```diff
@@ -833,7 +833,8 @@ default_object(Schema) ->
             kz_json:new()
     catch
         _Ex:_Err ->
-            lager:debug("exception getting schema default ~p : ~p", [_Ex, _Err]),
+            lager:debug("exception getting schema default ~p : ~p~nStackTrace: ~p",
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema.erl]: Revert this change. Add this after lager line instead:

```
ST = erlang:get_stacktrace(),
kz_util:log_stacktrace(ST),
```

File: applications/crossbar/priv/couchdb/schemas/system_config.crossbar.json [hunk 8]
```diff
@@ -148,6 +148,11 @@
             "description": "crossbar ssl workers",
             "type": "integer"
         },
+        "stability_level": {
```
Comment (icehess) [inline applications/crossbar/priv/couchdb/schemas/system_config.crossbar.json]: Add enum ti this key:

```
        "stability_level": {
            "default": "stable",
            "description": "crossbar stability_level",
            "enum": [
                "alpha",
                "beta",
                "stable"
            ],
            "type": "string"
        },
```

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 9]
```diff
@@ -115,3 +116,43 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+        InvalidKeys ->
+            InvalidKeysBin = binary:list_to_bin(lists:join(<<", ">>, InvalidKeys)),
+            ErrorMsg = <<"Disallowed properties -they've lower stability level than"
+                         " system's stability level" " (", SystemSL/binary, ")-: ",
+                         InvalidKeysBin/binary>>,
+            lager:debug("~s", [ErrorMsg]),
+            jesse_error:handle_data_invalid('external_error', ErrorMsg, State)
+    end.
+
+-spec stability_level_to_int(kz_term:ne_binary()) -> pos_integer().
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: Why you convert it to integer any way? why not just matching binary right away?

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 10]
```diff
@@ -115,3 +116,43 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+            State
+    end.
+
+check_params_stability_level(Properties, Value, State) ->
+    SystemSL = kapps_config:get_binary(<<"crossbar">>, <<"stability_level">>, <<"stable">>),
+    SystemSLInt = stability_level_to_int(SystemSL),
+    Fun = fun(Key, _Val, Acc) ->
+                  SearchKey = [Key, <<"stability_level">>],
+                  PropSL = kz_json:get_value(SearchKey, Properties, <<"stable">>),
+                  case stability_level_to_int(PropSL) < SystemSLInt of
```
Comment (icehess) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: Don't forget to change this line, since github thinks it changes. Look at my previous comment about it.

File: applications/crossbar/priv/api/swagger.json [hunk 11]
```diff
@@ -27613,6 +27613,16 @@
                     "description": "crossbar ssl workers",
                     "type": "integer"
                 },
+                "stability_level": {
+                    "default": "stable",
```
Comment (lazedo) [inline applications/crossbar/priv/api/swagger.json]: remove default

File: applications/crossbar/priv/couchdb/schemas/system_config.crossbar.json [hunk 12]
```diff
@@ -148,6 +148,16 @@
...
+            "default": "stable",
```
Comment (lazedo) [inline applications/crossbar/priv/couchdb/schemas/system_config.crossbar.json]: remove default

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 13]
```diff
@@ -115,3 +116,44 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+        [] ->
+            State; %% Not invalid keys found
+        InvalidKeys ->
+            InvalidKeysBin = binary:list_to_bin(lists:join(<<", ">>, InvalidKeys)),
+            ErrorMsg = <<"Disallowed properties -they've lower stability level than"
+                         " system's stability level" " (", SystemSL/binary, ")-: ",
+                         InvalidKeysBin/binary>>,
+            lager:debug("~s", [ErrorMsg]),
+            jesse_error:handle_data_invalid('external_error', ErrorMsg, State)
+    end.
```
Comment (lazedo) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: this seems wrong.
you should check the stability level for the value being checked, not for its children

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 14]
```diff
@@ -115,3 +116,44 @@ validate_attachment_oauth_doc_id(Value, State) ->
             lager:debug("~s", [ErrorMsg]),
             jesse_error:handle_data_invalid('external_error', ErrorMsg, State)
     end.
+
+maybe_check_params_stability_level(Value, Schema, State) ->
+    case {kz_json:get_value(<<"properties">>, Schema),
+          kz_json:get_value(<<"id">>, Schema),
+          kz_json:is_json_object(Value)}of
```
Comment (jamesaimonetti) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: Drop the `}` to newline and `of` to a newline after that:

```erlang
case {xyz
     ,pdq
     ,abc
     }
of
```

See [here](https://github.com/2600hz/kazoo/blob/master/core/kazoo_endpoint/src/kz_endpoint.erl#L146-L150) for an example

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 15]
```diff
@@ -115,3 +116,44 @@ validate_attachment_oauth_doc_id(Value, State) ->
...
+        {'undefined', _, _} ->
+            State; %% Skipping schema with out `properties' property
+        {Properties, Id, true} when Id /= <<"system_config.crossbar">> ->
+            check_params_stability_level(Properties, Value, State);
+        _ ->
+            State
+    end.
+
+check_params_stability_level(Properties, Value, State) ->
+    SystemSL = kapps_config:get_binary(<<"crossbar">>, <<"stability_level">>, <<"stable">>),
```
Comment (jamesaimonetti) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: Remove default here and the schemas should stop having it added

**Comment by jamesaimonetti** [comment]
My feeling is we should check the sysconfig value and, if `undefined`, should skip the checking of stability level all together. This would preserve backwards compat and allow folks to opt-in to stability level checking.

Should probably be per-account, as well, to allow having canary customers

**Comment by lazedo** [comment]
per-account should be a bit more complex, but global should be checked

**Comment by jamesaimonetti** [comment]
I agree that per-account introduces more work; probably not for this PR. But I know it will be an ask in the near future!

**Comment by lazedo** [comment]
@jamesaimonetti latest changes that get the param on crossbar side, will allow to easily add per-account

File: core/kazoo_schemas/src/kz_json_schema.erl [hunk 16]
```diff
@@ -41,19 +41,26 @@
 {'schema_loader_fun', fun((kz_term:ne_binary()) -> {'ok', kz_json:object()} | kz_json:object() | 'not_found')} |
 {'extra_validator', extra_validator()} |
 {'setter_fun', fun((kz_json:path(), kz_json:json_term(), kz_json:object()) -> kz_json:object())} |
-{'validator_options', validator_options()}.
+{'validator_options', validator_options()} |
+{'extra_validator_options', kz_term:proplist()}.
```
Comment (jamesaimonetti) [inline core/kazoo_schemas/src/kz_json_schema.erl]: Are there other options besides stability_level?

I'd like them enumerated here:

```erlang
-type extra_validator_option() :: {'stability_level', kz_term:ne_binary()}.
-type extra_validator_options() :: [extra_validator_option()].
```

This can be expanded as we go as well.

File: core/kazoo_schemas/src/kz_json_schema_extensions.erl [hunk 17]
```diff
@@ -5,16 +5,23 @@
...
         ]).
 
 -include_lib("kazoo_stdlib/include/kz_types.hrl").
 
 -define(INVALID_STORAGE_ATTACHMENT_REFERENCE(R), <<"invalid reference '", R/binary, "' to attachments">>).
 -define(INVALID_STORAGE_CONNECTION_REFERENCE(R), <<"invalid reference '", R/binary, "' to connections">>).
 
--spec extra_validator(jesse:json_term(), jesse_state:state()) -> jesse_state:state().
-extra_validator(Value, State) ->
+-spec extra_validator(jesse:json_term(), jesse_state:state(), kz_term:proplist()) -> jesse_state:state().
```
Comment (jamesaimonetti) [inline core/kazoo_schemas/src/kz_json_schema_extensions.erl]: reference specific type for Options instead of generic proplist.
